Source/WebCore/page/FrameView.cpp

```diff
@@ -2466,9 +2467,40 @@ void FrameView::textFragmentIndicatorTimerFired()
     TemporarySelectionChange selectionChange(document, { range }, { TemporarySelectionOption::DelegateMainFrameScroll, TemporarySelectionOption::RevealSelectionBounds, TemporarySelectionOption::UserTriggered });
     
     auto textIndicator = TextIndicator::createWithRange(range, { TextIndicatorOption::DoNotClipToVisibleRect }, WebCore::TextIndicatorPresentationTransition::Bounce);
-    if (textIndicator)
-        document.page()->chrome().client().setTextIndicator(textIndicator->data());
     
+    auto* page = frame().page();
+    
+    if (!page)
+        return;
+    
+    if (textIndicator) {
+        constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowVisibleChildFrameContentOnly };
+        
+        auto firstCharRange = resolveCharacterRange(range, { 0, 1 });
+        auto firstCharRect = RenderFlexibleBox::absoluteTextRects(firstCharRange);
+        
+        auto collapsedStart = SimpleRange(BoundaryPoint(range.startContainer(), range.startOffset()), BoundaryPoint(range.startContainer(), range.startOffset()));
+        
+        auto collapsedStartCharRects = RenderFlexibleBox::absoluteTextRects(collapsedStart);
+        
+        HitTestResult result;
+        result = page->mainFrame().eventHandler().hitTestResultAtPoint(LayoutPoint(collapsedStartCharRects.first().center()), hitType);
+        if (!intersects(range, *result.targetNode()))
+            return;
+
+        if (range.endOffset() > 1) {
+            auto lastCharRange = resolveCharacterRange(range, { range.endOffset() - 1, 1 });
+            auto lastCharRect = RenderFlexibleBox::absoluteTextRects(lastCharRange);
+            
+            auto collapsedEnd = SimpleRange(BoundaryPoint(range.endContainer(), range.endOffset()), BoundaryPoint(range.endContainer(), range.endOffset()));
+            auto collapsedEndCharRects = RenderFlexibleBox::absoluteTextRects(collapsedEnd);
+
+            result = page->mainFrame().eventHandler().hitTestResultAtPoint(LayoutPoint(lastCharRect.first().center()), hitType);
+            if (!intersects(range, *result.targetNode()))
+                return;
+        }
+        document.page()->chrome().client().setTextIndicator(textIndicator->data());
+    }
     cancelScheduledTextFragmentIndicatorTimer();
```

Comment:
Move this up before the if, invert the if, and make it an early return?

Reply:
I didn't do an early return because we do not want to cancel the flash if we have only searched for a character.